docs/ds20.md

```diff
@@ -0,0 +1,111 @@
+We do need to use the `BOS` *“Binary Object Store”* descriptor, which is only supported for devices using USB 2.1 and newer, although that’s probably the vast majority of devices now.
+Usually the BOS descriptor is used for Wireless USB details, USB 2.0 extensions, SuperSpeed USB connection details and a Container ID.
+
+Usefully, the BOS descriptor does give us is the ability to use the platform capability descriptor, which is `bDevCapabilityType=0x05`. For UUID `D8DD60DF-4589-4CC7-9CD2-659D9E648A9F`, this is identified as a structured blob of data Windows usually uses to put workarounds like the suspend mode of the device and that kind of thing.
+
+We could certainly create a new `bDevCapabilityType` which would allow us to store a binary blob (e.g. `Plugin=foobarbaz\nFlags=QuirkValueHere\n`) but that would be out-of-specification and difficult for vendors to implement.
+
+## Proposal
+
+Devices create a new platform device capability BOS descriptor with `bDevCapabilityType=0x05` and a fwupd-specific UUID, specifically `010aec63-f574-52cd-9dda-2852550d94f0`.
+Using this custom UUID will ensure that Microsoft Windows does not try to parse the capability descriptor as a *Microsoft OS 2.0 descriptor set*.
+
+## Implementation
+
+Create a BOS DS20 descriptor such as:
+
+    1C 10 05 00 63 ec 0a 01 74 f5 cd 52 9d da 28 52 55 0d 94 f0 05 08 01 00 20 00 2a 00
+    |  |  |  |  \---------------------------------------------/ \--------/  \---/ |  |
+    |  |  |  |                ^-PlatformCapabilityUUID            |   wLength-^   |  |
+    |  |  |  ^-bReserved                                 dwVersion^   bVendorCode-^  |
+    |  |  ^-bDevCapability                                                bAltEnumCmd^
+    |  ^-bDescriptorType
+    ^-bLength
+
+The `dwVersion` encoded here is fwupd `1.8.5`, which is also the first version that supports BOS DS20 descriptors.
+A `dwVersion` of 0x0 will match any fwupd version, but should probably not be used unless required.
+
+The BOS descriptors are sorted by the requester and can appear in any order.
+The descriptor with the highest `dwVersion` that is not newer than the current running daemon version is used.
+This allows devices to set different quirks depending on the fwupd version, although in practice this should not be required.
+The `bAltEnumCmd` parameter **must** be zero.
+The `PlatformCapabilityUUID` **must** be `010aec63-f574-52cd-9dda-2852550d94f0`.
+
+Then allow the device to reply to a USB control request with the following parameters:
+
+    transfer-direction: device-to-host
+    request-type: vendor
+    recipient: device
+    bRequest: {value of bVendorCode in the BOS descriptor)
+    wValue: 0x00
+    wIndex: 0x07
+    wLength: {value of wLength in the BOS descriptor)
+
+The device should return something like:
+
+    50 6c 75 67 69 6e 3d 64 66 75 0a 49 63 6f 6e 3d 63 6f 6d 70 75 74 65 72 0a 00 00 00 00 00 00 00
+
+...which is the UTF-8 quirk data, e.g.
```

Comment:
This means _USB devices_ have a new attack surface area of the quirk parser which otherwise was just reading files distributed with fwupd.

I think we'll want to be doing something in the code to ensure that the plugins that can be used are only USB plugins.

I also think that there needs to be a required comparison that a USB device's `idVendor` can only be used with _certain_  USB plugins.

Reply:
I've done the UTF-8 thing. I've also reviewed all the quick parser vfuncs, and we're doing a really good job with things like `fu_strtoull` rather than doing unsafe parsing of value. I'm not super worried about it tbh.

Reply:
> I've done the UTF-8 thing. I've also reviewed all the quick parser vfuncs, and we're doing a really good job with things like `fu_strtoull` rather than doing unsafe parsing of value. I'm not super worried about it tbh.

OK sounds good.  Hopefully the fuzzers can catch anything remaining then.

Reply:
Fuzzers, hmm, indeed. Wilfix.

Reply:
Sounds good!